File: wire-ios-sync-engine/Tests/Source/Integration/NotificationObservers.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 

@import WireDataModel;

#import "NotificationObservers.h"

@interface ChangeObserver ()
@end



@implementation ChangeObserver

- (instancetype)init
{
    self = [super init];
    if (self) {
        _notifications = [[NSMutableArray alloc] init];
        [self startObservering];
    }
    return self;
}

- (void)dealloc
{
    [self stopObservering];
}

- (void)clearNotifications;
{
    [self.notifications removeAllObjects];
}

- (void)startObservering;
{
}

- (void)stopObservering;
{
}

- (void)tearDown;
{
}

@end

@interface ConversationChangeObserver ()
@property (nonatomic, weak) ZMConversation *conversation;
@property (nonatomic) id token;
@end


@implementation ConversationChangeObserver

- (instancetype)initWithConversation:(ZMConversation *)conversation
{
    self = [super init];
    if(self) {
        self.conversation = conversation;
        self.token = [ConversationChangeInfo addObserver:self forConversation:conversation];
    }
    return self;
}

- (void)conversationDidChange:(ConversationChangeInfo *)note;
{
    [self.notifications addObject:note];
    if (self.notificationCallback) {
        self.notificationCallback(note);
    }
}


@end



@interface ConversationListChangeObserver ()
@property (nonatomic, weak) ZMConversationList *conversationList;
@property (nonatomic) id token;
@end

@implementation ConversationListChangeObserver

ZM_EMPTY_ASSERTING_INIT()

- (instancetype)initWithConversationList:(ZMConversationList *)conversationList;
{
    self = [super init];
    if(self) {
        self.conversationList = conversationList;
        self.conversationChangeInfos = [NSMutableArray array];
        self.token = [ConversationListChangeInfo addObserver:self forList:conversationList managedObjectContext:conversationList.managedObjectContext];
    }
    return self;
}

- (void)conversationListDidChange:(ConversationListChangeInfo *)note;
{
    [self.notifications addObject:note];
    if (self.notificationCallback) {
        self.notificationCallback(note);
    }
}

- (void)conversationInsideList:(ZMConversationList *)list didChange:(ConversationChangeInfo *)changeInfo;
{
    NOT_USED(list);
    [self.conversationChangeInfos addObject:changeInfo];
}

@end



@interface UserChangeObserver ()
@property (nonatomic, weak) id<UserType> user;
@property (nonatomic) id token;
@end

@implementation UserChangeObserver

- (instancetype)initWithUser:(ZMUser *)user
{
    return [self initWithUser:user managedObjectContext:user.managedObjectContext];
}

- (instancetype)initWithUser:(id<UserType>)user managedObjectContext:(NSManagedObjectContext *)managedObjectContext
{
    self = [super init];
    if(self) {
        self.user = user;
        self.token = [UserChangeInfo addObserver:self forUser:user inManagedObjectContext:managedObjectContext];
    }
    return self;
}

- (void)startObservering;
{
}

- (void)stopObservering
{
}

- (void)userDidChange:(UserChangeInfo *)info;
{
    [self.notifications addObject:info];
    if (self.notificationCallback) {
        self.notificationCallback(info);
    }
}


@end



@interface MessageChangeObserver ()
@property (nonatomic, weak) ZMMessage *message;
@property (nonatomic) id token;
@end

@implementation MessageChangeObserver

- (instancetype)initWithMessage:(ZMMessage *)message
{
    self = [super init];
    if(self) {
        self.message = message;
        self.token = [MessageChangeInfo addObserver:self forMessage:message managedObjectContext:message.managedObjectContext];
    }
    return self;
}

- (void)messageDidChange:(MessageChangeInfo *)changeInfo
{
    [self.notifications addObject:changeInfo];
}

@end

File: wire-ios-data-model/Tests/Source/Helper/NotificationObservers.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation

// MARK: - ObserverType

protocol ObserverType: NSObjectProtocol {
    associatedtype ChangeInfo: ObjectChangeInfo
    var notifications: [ChangeInfo] { get set }
}

extension ObserverType {
    func clearNotifications() {
        notifications = []
    }
}

// MARK: - MockUserObserver

final class MockUserObserver: UserObserving {
    var notifications = [UserChangeInfo]()

    func clearNotifications() {
        notifications = []
    }

    func userDidChange(_ changeInfo: UserChangeInfo) {
        notifications.append(changeInfo)
    }
}

// MARK: - MessageObserver

class MessageObserver: NSObject, ZMMessageObserver {
    var token: NSObjectProtocol?

    override init() {}

    init(message: ZMMessage) {
        super.init()
        self.token = MessageChangeInfo.add(
            observer: self,
            for: message,
            managedObjectContext: message.managedObjectContext!
        )
    }

    var notifications: [MessageChangeInfo] = []

    func messageDidChange(_ changeInfo: MessageChangeInfo) {
        notifications.append(changeInfo)
    }
}

// MARK: - NewUnreadMessageObserver

class NewUnreadMessageObserver: NSObject, ZMNewUnreadMessagesObserver {
    var token: NSObjectProtocol?
    var notifications: [NewUnreadMessagesChangeInfo] = []

    override init() {}

    init(context: NSManagedObjectContext) {
        super.init()
        self.token = NewUnreadMessagesChangeInfo.add(observer: self, managedObjectContext: context)
    }

    func didReceiveNewUnreadMessages(_ changeInfo: NewUnreadMessagesChangeInfo) {
        notifications.append(changeInfo)
    }
}

// MARK: - ConversationObserver

final class ConversationObserver: NSObject, ZMConversationObserver {
    var token: NSObjectProtocol?

    func clearNotifications() {
        notifications = []
    }

    override init() {}

    init(conversation: ZMConversation) {
        super.init()
        self.token = ConversationChangeInfo.add(observer: self, for: conversation)
    }

    var notifications = [ConversationChangeInfo]()

    func conversationDidChange(_ changeInfo: ConversationChangeInfo) {
        notifications.append(changeInfo)
    }
}

// MARK: - ConversationListChangeObserver

@objcMembers
class ConversationListChangeObserver: NSObject, ZMConversationListObserver {
    public var notifications = [ConversationListChangeInfo]()
    public var observerCallback: ((ConversationListChangeInfo) -> Void)?
    unowned var conversationList: ConversationList
    var token: NSObjectProtocol?

    init(conversationList: ConversationList, managedObjectContext: NSManagedObjectContext) {
        self.conversationList = conversationList
        super.init()
        self.token = ConversationListChangeInfo.addListObserver(
            self,
            for: conversationList,
            managedObjectContext: managedObjectContext
        )
    }

    func conversationListDidChange(_ changeInfo: ConversationListChangeInfo) {
        notifications.append(changeInfo)
        if let callBack = observerCallback {
            callBack(changeInfo)
        }
    }
}
